cranelift/filetests/src/runtest_environment.rs

```diff
@@ -0,0 +1,111 @@
+
+/// Stores info about the expected environment for a test function.
+#[derive(Debug, Clone)]
+pub struct RuntestEnvironment {
+    pub heaps: Vec<HeapCommand>,
+}
+
+impl RuntestEnvironment {
+    /// Parse the environment from a set of comments
+    pub fn parse(comments: &[Comment]) -> anyhow::Result<Self> {
+        let mut env = RuntestEnvironment { heaps: Vec::new() };
+
+        for comment in comments.iter() {
+            if let Some(heap_command) = parse_heap_command(comment.text)? {
+                let heap_index = env.heaps.len() as u64;
+                let expected_ptr = heap_index * 16;
+                if Some(expected_ptr) != heap_command.ptr_offset.map(|p| p.into()) {
+                    return Err(anyhow!(
+                        "Invalid ptr offset, expected vmctx+{}",
+                        expected_ptr
+                    ));
+                }
+
+                let expected_bound = (heap_index * 16) + 8;
+                if Some(expected_bound) != heap_command.bound_offset.map(|p| p.into()) {
+                    return Err(anyhow!(
+                        "Invalid bound offset, expected vmctx+{}",
+                        expected_bound
+                    ));
+                }
+
+                env.heaps.push(heap_command);
+            };
+        }
+
+        Ok(env)
+    }
+
+    pub fn is_active(&self) -> bool {
+        !self.heaps.is_empty()
+    }
+
+    /// Allocates a struct to be injected into the test.
+    pub fn runtime_struct(&self) -> RuntestContext {
+        RuntestContext::new(&self)
+    }
+}
+
+type HeapMemory = Vec<u8>;
+
+/// A struct that provides info about the environment to the test
+#[derive(Debug, Clone)]
+pub struct RuntestContext {
+    /// Store the heap memory alongside the context info so that we don't accidentally deallocate
+    /// it too early.
+    heaps: Vec<HeapMemory>,
+
+    /// This is the actual struct that gets passed into the `vmctx`  argument of the tests.
+    /// It has a specific memory layout that all tests agree with.
+    ///
+    /// Currently we only have to store heap info, so we store the heap start and end addresses in
+    /// a 64 bit slot for each heap.
+    ///
+    /// ┌────────────┐
+    /// │heap0: start│
+    /// ├────────────┤
+    /// │heap0: end  │
+    /// ├────────────┤
+    /// │heap1: start│
+    /// ├────────────┤
+    /// │heap1: end  │
+    /// ├────────────┤
+    /// │etc...      │
+    /// └────────────┘
+    context_struct: Vec<u64>,
+}
+
+impl RuntestContext {
+    pub fn new(env: &RuntestEnvironment) -> Self {
+        let heaps: Vec<HeapMemory> = env
+            .heaps
+            .iter()
+            .map(|cmd| {
+                let size: u64 = cmd.size.into();
+                vec![0u8; size as usize]
+            })
+            .collect();
+
+        let context_struct = heaps
+            .iter()
+            .flat_map(|heap| [heap.as_ptr(), unsafe { heap.as_ptr().add(heap.len()) }])
```

Comment:
1. Should this be handling overflow?
2. And what changes here if/when #3187 is merged?
3. And are we going to reuse this struct when we fuzz CLIF + heaps?

Reply:
> Should this be handling overflow?

According to the rust docs, [this should always be safe](https://doc.rust-lang.org/std/primitive.pointer.html#safety-5). But even in the event where this goes wrong I would expect the tests to fail with a SIGSEGV. It might be better to change to the `wrapping_add` variant of this function.

> And what changes here if/when cranelift: Add stack support to the interpreter with virtual addresses #3187 is merged?

We won't be able to do this the same way with the interpreter, it will probably need some refactoring to handle the separate address space. Off the top of my head, I think we can do something like the following:

- Allocate the heaps as usual
- Register each heap with the interpreter
- Request the heap pointers from the interpreter
- Build the `context_struct` with those pointers.

> And are we going to reuse this struct when we fuzz CLIF + heaps?

I hadn't thought about that, but it actually is an easy way to pass all of this info while fuzzing to the interpreter and the actual compiled target.

For fuzzing I'd like to relax the requirement that `vmctx` is the first argument, but that shouldn't be an issue, since we only check for that in the parser.

We may also need to coordinate the heaps with the generated `global_values` in each function. Otherwise I think the fuzzer may have issues generating the correct offsets to get valid heap accesses.

I think we should be OK reusing this for fuzzing with the changes to generate the valid pointers for the interpreter.